superset/tasks/async_queries.py

```diff
@@ -0,0 +1,111 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import logging
+from typing import Any, cast, Dict, Optional
+
+from flask import current_app
+
+from superset import app
+from superset.exceptions import SupersetVizException
+from superset.extensions import async_query_manager, cache_manager, celery_app
+from superset.utils.cache import generate_cache_key, set_and_log_cache
+from superset.views.utils import get_datasource_info, get_viz
+
+logger = logging.getLogger(__name__)
+query_timeout = current_app.config[
+    "SQLLAB_ASYNC_TIME_LIMIT_SEC"
+]  # TODO: new config key
+
+
+@celery_app.task(name="load_chart_data_into_cache", soft_time_limit=query_timeout)
+def load_chart_data_into_cache(
+    job_metadata: Dict[str, Any], form_data: Dict[str, Any],
+) -> None:
+    from superset.charts.commands.data import (
+        ChartDataCommand,
+    )  # load here due to circular imports
+
+    with app.app_context():  # type: ignore
+        try:
+            command = ChartDataCommand()
+            command.set_query_context(form_data)
+            result = command.run(cache=True)
+            cache_key = result["cache_key"]
+            result_url = f"/api/v1/chart/data/{cache_key}"
+            async_query_manager.update_job(
+                job_metadata, async_query_manager.STATUS_DONE, result_url=result_url,
+            )
+        except Exception as exc:
+            # TODO: QueryContext should support SIP-40 style errors
+            error = exc.message if hasattr(exc, "message") else str(exc)  # type: ignore # pylint: disable=no-member
+            errors = [{"message": error}]
+            async_query_manager.update_job(
+                job_metadata, async_query_manager.STATUS_ERROR, errors=errors
+            )
+            raise exc
+
+        return None
+
+
+@celery_app.task(name="load_explore_json_into_cache", soft_time_limit=query_timeout)
+def load_explore_json_into_cache(
```

Comment:
Background job for running async chart queries in the context of the legacy `/superset/explore_json` API